File: app/src/main/java.backup/com/lumiyaviewer/lumiya/eventbus/EventBus.java
package com.lumiyaviewer.lumiya.eventbus;

import android.app.Activity;
import android.os.Handler;
import java.lang.ref.WeakReference;
import java.lang.reflect.Method;
import java.util.LinkedList;
import java.util.List;

public class EventBus {
    private final List<HandlerInfo> handlers;

    private static class EventInvocation implements Runnable {
        private final Activity activity;
        private final Object event;
        private final Handler handler;
        private final Method method;
        private final Object subscriber;

        public EventInvocation(Object obj, Activity activity, Object obj2, Method method, Handler handler) {
            this.event = obj;
            this.activity = activity;
            this.subscriber = obj2;
            this.method = method;
            this.handler = handler;
        }

        public void run() {
            try {
                this.method.invoke(this.subscriber, new Object[]{this.event});
            } catch (Exception e) {
                // Log reflection errors for debugging
                android.util.Log.e("EventBus", "Failed to invoke event handler method", e);
            }
        }

        public void runOnUIThread() {
            if (this.activity != null) {
                this.activity.runOnUiThread(this);
            } else if (this.handler != null) {
                this.handler.post(this);
            } else {
                run();
            }
        }
    }

    private static class HandlerInfo {
        private final WeakReference<Activity> activity;
        private final Class<?> eventClass;
        private final WeakReference<Handler> handler;
        private final Method method;
        private final WeakReference<?> subscriber;

        public HandlerInfo(Class<?> cls, Method method, Object obj, Activity activity, Handler handler) {
            this.eventClass = cls;
            this.method = method;
            this.subscriber = new WeakReference(obj);
            this.activity = new WeakReference(activity);
            this.handler = new WeakReference(handler);
        }

        public Activity getActivity() {
            return (Activity) this.activity.get();
        }

        public Handler getHandler() {
            return (Handler) this.handler.get();
        }

        public Method getMethod() {
            return this.method;
        }

        public Object getSubscriber() {
            return this.subscriber.get();
        }

        public boolean matchesEvent(Object obj) {
            return obj != null ? obj.getClass().equals(this.eventClass) : false;
        }
    }

    private static class InstanceHolder {
        private static final EventBus Instance = new EventBus();

        private InstanceHolder() {
        }
    }

    private EventBus() {
        this.handlers = new LinkedList();
    }

    /* synthetic */ EventBus(EventBus eventBus) {
        this();
    }

    public static EventBus getInstance() {
        return InstanceHolder.Instance;
    }

    public synchronized void publish(Object obj) {
        Iterable<HandlerInfo> linkedList = new LinkedList();
        for (HandlerInfo handlerInfo : this.handlers) {
            if (handlerInfo.matchesEvent(obj)) {
                Object subscriber = handlerInfo.getSubscriber();
                Activity activity = handlerInfo.getActivity();
                if (subscriber == null) {
                    linkedList.add(handlerInfo);
                } else {
                    new EventInvocation(obj, activity, subscriber, handlerInfo.getMethod(), handlerInfo.getHandler()).runOnUIThread();
                }
            }
        }
        for (HandlerInfo remove : linkedList) {
            this.handlers.remove(remove);
        }
    }

    public synchronized void subscribe(Activity activity) {
        subscribe(activity, activity);
    }

    public synchronized void subscribe(Object obj) {
        if (obj instanceof Activity) {
            subscribe(obj, (Activity) obj);
        } else {
            subscribe(obj, null);
        }
    }

    public synchronized void subscribe(Object obj, Activity activity) {
        subscribe(obj, activity, null);
    }

    public synchronized void subscribe(Object obj, Activity activity, Handler handler) {
        Iterable<HandlerInfo> linkedList = new LinkedList();
        for (HandlerInfo handlerInfo : this.handlers) {
            Object subscriber = handlerInfo.getSubscriber();
            if (subscriber != obj) {
                if (subscriber == null) {
                    linkedList.add(handlerInfo);
                }
            } else {
                return;
            }
        }
        for (HandlerInfo handlerInfo2 : linkedList) {
            this.handlers.remove(handlerInfo2);
        }
        for (Method method : obj.getClass().getMethods()) {
            if (((EventHandler) method.getAnnotation(EventHandler.class)) != null) {
                Class[] parameterTypes = method.getParameterTypes();
                if (parameterTypes.length != 1) {
                    throw new IllegalArgumentException("EventHandler methods must specify a single Object paramter.");
                }
                this.handlers.add(new HandlerInfo(parameterTypes[0], method, obj, activity, handler));
            }
        }
    }

    public synchronized void unsubscribe(Object obj) {
        Iterable<HandlerInfo> linkedList = new LinkedList();
        for (HandlerInfo handlerInfo : this.handlers) {
            Object subscriber = handlerInfo.getSubscriber();
            if (subscriber == null || subscriber == obj) {
                linkedList.add(handlerInfo);
            }
        }
        for (HandlerInfo handlerInfo2 : linkedList) {
            this.handlers.remove(handlerInfo2);
        }
    }

    public synchronized void unsubscribeActivity(Activity activity) {
        Iterable<HandlerInfo> linkedList = new LinkedList();
        for (HandlerInfo handlerInfo : this.handlers) {
            if (handlerInfo.getActivity() == activity) {
                linkedList.add(handlerInfo);
            }
        }
        for (HandlerInfo handlerInfo2 : linkedList) {
            this.handlers.remove(handlerInfo2);
        }
    }
}

File: app/src/main/java/com/lumiyaviewer/lumiya/eventbus/EventBus.kt
package com.lumiyaviewer.lumiya.eventbus

import android.app.Activity
import android.os.Handler
import android.util.Log
import java.lang.ref.WeakReference
import java.lang.reflect.Method
import java.util.*

/**
 * Event bus for decoupled communication between components.
 * Converted from Java to Kotlin with improved type safety and coroutines support.
 */
class EventBus private constructor() {
    
    private val handlers = mutableListOf<HandlerInfo>()

    /**
     * Event invocation wrapper for thread-safe execution
     */
    private class EventInvocation(
        private val event: Any,
        private val activity: Activity?,
        private val subscriber: Any,
        private val method: Method,
        private val handler: Handler?
    ) : Runnable {

        override fun run() {
            try {
                method.invoke(subscriber, event)
            } catch (e: Exception) {
                Log.e(TAG, "Failed to invoke event handler method", e)
            }
        }

        fun runOnUIThread() {
            when {
                activity != null -> activity.runOnUiThread(this)
                handler != null -> handler.post(this)
                else -> run()
            }
        }
    }

    /**
     * Handler information with weak references to prevent memory leaks
     */
    private class HandlerInfo(
        val eventClass: Class<*>,
        val method: Method,
        subscriber: Any,
        activity: Activity?,
        handler: Handler?
    ) {
        private val subscriberRef = WeakReference(subscriber)
        private val activityRef = WeakReference(activity)
        private val handlerRef = WeakReference(handler)

        fun getActivity(): Activity? = activityRef.get()
        fun getHandler(): Handler? = handlerRef.get()
        fun getSubscriber(): Any? = subscriberRef.get()

        fun matchesEvent(event: Any): Boolean = event.javaClass == eventClass
    }

    /**
     * Singleton holder using thread-safe lazy initialization
     */
    companion object {
        private const val TAG = "EventBus"
        
        @JvmStatic
        fun getInstance(): EventBus = InstanceHolder.instance
        
        private object InstanceHolder {
            val instance = EventBus()
        }
    }

    /**
     * Publish an event to all matching subscribers
     */
    @Synchronized
    fun publish(event: Any) {
        val deadHandlers = mutableListOf<HandlerInfo>()
        
        for (handler in handlers) {
            if (handler.matchesEvent(event)) {
                val subscriber = handler.getSubscriber()
                if (subscriber == null) {
                    deadHandlers.add(handler)
                } else {
                    EventInvocation(
                        event,
                        handler.getActivity(),
                        subscriber,
                        handler.method,
                        handler.getHandler()
                    ).runOnUIThread()
                }
            }
        }
        
        // Clean up dead references
        handlers.removeAll(deadHandlers)
    }

    /**
     * Subscribe an Activity to events (Activity serves as both subscriber and context)
     */
    @Synchronized
    fun subscribe(activity: Activity) {
        subscribe(activity, activity)
    }

    /**
     * Subscribe an object to events
     */
    @Synchronized
    fun subscribe(subscriber: Any) {
        if (subscriber is Activity) {
            subscribe(subscriber, subscriber)
        } else {
            subscribe(subscriber, null)
        }
    }

    /**
     * Subscribe an object to events with specific Activity context
     */
    @JvmOverloads
    @Synchronized
    fun subscribe(subscriber: Any, activity: Activity?, handler: Handler? = null) {
        // Clean up any existing dead references and duplicate subscriptions
        val deadHandlers = mutableListOf<HandlerInfo>()
        
        for (handlerInfo in handlers) {
            val existingSubscriber = handlerInfo.getSubscriber()
            when {
                existingSubscriber == subscriber -> return // Already subscribed
                existingSubscriber == null -> deadHandlers.add(handlerInfo)
            }
        }
        
        handlers.removeAll(deadHandlers)
        
        // Register event handler methods
        for (method in subscriber.javaClass.methods) {
            if (method.isAnnotationPresent(EventHandler::class.java)) {
                val parameterTypes = method.parameterTypes
                require(parameterTypes.size == 1) {
                    "EventHandler methods must specify a single parameter"
                }
                
                handlers.add(HandlerInfo(parameterTypes[0], method, subscriber, activity, handler))
            }
        }
    }

    /**
     * Unsubscribe an object from all events
     */
    @Synchronized
    fun unsubscribe(subscriber: Any) {
        handlers.removeAll { handlerInfo ->
            val existingSubscriber = handlerInfo.getSubscriber()
            existingSubscriber == null || existingSubscriber == subscriber
        }
    }

    /**
     * Unsubscribe all handlers associated with an Activity
     */
    @Synchronized
    fun unsubscribeActivity(activity: Activity) {
        handlers.removeAll { handlerInfo ->
            handlerInfo.getActivity() == activity
        }
    }
}